j2v/generation/generator.py

```diff
@@ -25,81 +25,93 @@ def clean(self):
-                relative_path = current_path + ":" + doublequote(key)
-                self.collect_all_paths(value, relative_path, current_view, root_view, key)
-            elif is_non_empty_1D_list(value):
-                new_view_name = self.get_full_path_str(current_view, current_path, key)
-                sample_element = value[0]
-                if is_dict(sample_element):
-                    self.add_explore_join(new_view_name, current_view, key, current_path)
-                    self.collect_all_paths(current_dict=sample_element,
-                                           current_path=ELEMENT_ACCESS_STR,
-                                           current_view=new_view_name,
-                                           root_view=current_view)
-
-                elif is_primitive(sample_element):
-                    self.add_explore_join(new_view_name, current_view, key, current_path)
-                    self.add_dimension("", new_view_name, ELEMENT_ACCESS_STR, sample_element, None,
-                                       primitive_array=True)
-
-    def get_full_path_str(self, current_view, current_path, key):
-        """
 
+        if is_primitive(data_object) or data_object is None:
+            self.add_dimension(current_path, current_view, data_object_key or ELEMENT_ACCESS_STR, data_object,
+                               parent_object_key)
+        elif is_dict(data_object):
+            str_keys_only = filter(lambda item: type(item[0]) == str, data_object.items())
+            primitives_first_items = sorted(str_keys_only, key=lambda x: self.sortation_order(x[1]))
+            for key, value in primitives_first_items:
+                self.collect_all_paths(data_object=value, current_path=current_path, current_view=current_view,
+                                       data_object_key=key,
+                                       parent_object_key=data_object_key)
+        elif is_non_empty_1D_list(data_object):
+            new_view_name = self.get_new_view_name(current_view, current_path)
+            sample_element = data_object[0]
+            self.add_explore_join(new_view_name, current_view, current_path)
+            self.collect_all_paths(data_object=sample_element,
+                                   current_path=current_path,
+                                   current_view=new_view_name)
+
+    @staticmethod
+    def sortation_order(value):
+        def depth(x):
+            if type(x) is dict and x:
+                return 1 + max(depth(x[a]) for a in x)
+            if type(x) is list and x:
+                return 1 + max(depth(a) for a in x)
+            return 0
+
+        # put primitives at the first place, since we will not have altrenatives in case of duplicated dim names
+        # the deepest dicts should be processed at the end
+        if is_primitive(value) or value is None:
+            return -1
+        return depth(value)
+
+    def get_new_view_name(self, current_view, current_path):
```

Comment:
Why is this called `get_new_view_name`? What does the `new` stand for?

Reply:
Renamed.